tests/testthat/test-xgboost-deploy.R

```diff
@@ -73,3 +79,226 @@ test_that("Top categories are correctly parsed", {
+  capture.output(boost$run())
+  
+  # Deploy set contains new factor levels. These should be set to NA and
+  # imputed
+  dfDeploy <- data.frame(id = c(1789, 1790, 1791),
+                         color = c("golden", NA, "yellow"),
+                         shape = c("unbending", NA, "straight"))
+  
+  p2 <- SupervisedModelDeploymentParams$new()
+  p2$type <- "multiclass"
+  p2$df <- dfDeploy
+  p2$grainCol <- "id"
+  p2$predictedCol <- "object"
+  p2$impute <- TRUE
+  p2$debug <- F
+  # deploy model
+  xNew <- capture.output(boostD <- XGBoostDeployment$new(p2))
+  # Check that warning is triggered
+  expect_warning(xDeploy <- capture.output(boostD$deploy()),
+                 paste("New categorical variable levels were found:\n",
+                       " -  color : golden\n",
+                       " -  shape : unbending\n",
+                       "These values have been set to NA.",
+                       sep = ""))
+  
+  # Get output dataframe for csv or SQL
+  xDf <- capture.output(outDf <- boostD$getOutDf())
+  
+  # Check the predictions of the first 2 rows match
+  expect_equal(outDf[1, 5], outDf[2, 5])
+  expect_equal(outDf[1, 6], outDf[2, 6])
+  expect_equal(outDf[1, 7], outDf[2, 7])
+})
+
+test_that("Factors and characters give the same results", {
+  # train model on dataset with factors
+  p <- getToyXGBDevelopParams(asFactor = T)
+  capture.output(boost <- XGBoostDevelopment$new(p))
+  capture.output(boost$run())
+  
+  # deploy set also uses factors
+  dfDeployFact <- data.frame(id = c(1776, 1767),
+                             color = c("yellow", "green"),
+                             shape = c("curved", "round"),
+                             stringsAsFactors = T)
+  p2 <- SupervisedModelDeploymentParams$new()
+  p2$type <- "multiclass"
+  p2$df <- dfDeployFact
+  p2$grainCol <- "id"
+  p2$predictedCol <- "object"
+  p2$impute <- TRUE
+  p2$debug <- F
+  
+  # deploy model and get output dataframe for factors
+  xNewFact <- capture.output(boostDFact <- XGBoostDeployment$new(p2))
+  xDeployFact <- capture.output(boostDFact$deploy())
+  xDfFact <- capture.output(outDfFact <- boostDFact$getOutDf())
+  
+  # develop model on same dataset, but with characters instead of factors
+  p <- getToyXGBDevelopParams(asFactor = F)
+  capture.output(boost <- XGBoostDevelopment$new(p))
+  capture.output(boost$run())
+  
+  # deploy set also uses characters
+  dfDeployChar <- data.frame(id = c(1776, 1767),
+                             color = c("yellow", "green"),
+                             shape = c("curved", "round"),
+                             stringsAsFactors = F)
+  p2$df <- dfDeployChar
+  
+  # deploy model and get output dataframe for characters
+  xNewChar <- capture.output(boostDChar <- XGBoostDeployment$new(p2))
+  xDeployChar <- capture.output(boostDChar$deploy())
+  xDfChar <- capture.output(outDfChar <- boostDChar$getOutDf())
+  
+  # Check that we get the same results when using factors/characters
```

Comment:
Actually, all of them are great. If these pass, I'm happy with xgb. Until we find more bugs of course 😄